### PR TITLE
ARROW-12944: [C++] String capitalize kernel

### PR DESCRIPTION
**Pull request by edponce**
This PR adds scalar compute functions for string capitalization, namely "ascii_capitalize" and "utf8_capitalize".

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 1]
```diff
@@ -395,6 +395,13 @@ TYPED_TEST(TestStringKernels, AsciiLower) {
                    "[\"aaazzæÆ&\", null, \"\", \"bbb\"]");
 }
 
+TYPED_TEST(TestStringKernels, AsciiCapitalize) {
+  this->CheckUnary("ascii_capitalize", "[]", this->type(), "[]");
+  this->CheckUnary("ascii_capitalize",
+                   "[\"aAazZæÆ&\", null, \"\", \"bBB\", \"one word\"]", this->type(),
+                   "[\"AAazZæÆ&\", null, \"\", \"BBB\", \"One word\"]");
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:402]: Can you test a bit more thoroughly?
For example in Python:
```python
>>> "hEllo world".capitalize()
'Hello world'
```

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 2]
```diff
@@ -4018,6 +4036,13 @@ const FunctionDoc ascii_lower_doc(
      "non-ASCII characters, use \"utf8_lower\" instead."),
     {"strings"});
 
+const FunctionDoc ascii_capitalize_doc(
+    "Capilatize the first character of ASCII input",
+    ("For each string in `strings`, return a capitalized version.\n\n"
+     "This function assumes the input is fully ASCII.  If it may contain\n"
+     "non-ASCII characters, use \"utf8_capitalize\" instead."),
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:4043]: I don't see utf8_capitalize in this PR. Is this an oversight?

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 3]
```diff
@@ -443,6 +443,24 @@ struct AsciiLower {
...
+  int64_t Transform(const uint8_t* input, int64_t input_string_ncodeunits,
+                    uint8_t* output) {
+    if (input_string_ncodeunits > 0) {
+      output[0] = ascii_toupper(input[0]);
+      std::memcpy(output + 1, input + 1, input_string_ncodeunits - 1);
+    }
+    return input_string_ncodeunits;
+  }
+
+  Status InvalidStatus() override {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:456]: It's not clear to me why this function is being overriden if the function always returns successfully.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 2 above)
Comment (edponce) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:4043]: Working on it locally, I should have made a draft PR.

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 3 above)
Comment (edponce) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:456]: An artifact of copy-paste from another compute function...removed.

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 4]
```diff
@@ -4028,17 +4087,14 @@ const FunctionDoc utf8_lower_doc(
...
 
-const FunctionDoc ascii_reverse_doc(
-    "Reverse ASCII input",
-    ("For each ASCII string in `strings`, return a reversed version.\n\n"
-     "This function assumes the input is fully ASCII.  If it may contain\n"
-     "non-ASCII characters, use \"utf8_reverse\" instead."),
-    {"strings"});
+const FunctionDoc utf8_capitalize_doc(
+    "Capitalize the first codepoint of UTF8 input",
+    ("For each UTF8 string in `strings`, return a capitalized version."), {"strings"});
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:4092]: Nit, but no need to specify "UTF8". The input is of string type, so UTF8 by definition.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Rebased, will merge if green.

**Comment by edponce** [comment]
@pitrou I pushed a change after you rebased, so I will fix conflicts and re-push.

**Comment by pitrou** [comment]
Hmm, another commit was lost then...

**Comment by pitrou** [comment]
Recommend starting from https://github.com/apache/arrow/commit/9d7f298bf1515c0014b2cd138c475e7bc1deda87 and then applying the buffer overflow fix.

**Comment by edponce** [comment]
Some tests invoke the incorrect kernel (ASCII test uses `utf8_capitalize` and viceversa). These "typos" are fixed in https://github.com/apache/arrow/pull/10869

**Pull request closed without merging**